File: savings_algorithm.py
"""
Clarke-Wright Savings Algorithm for Vehicle Routing Problem (VRP)

This classical heuristic starts with individual routes from depot to each customer
and then combines routes based on savings calculations to minimize total distance.
"""

import math
from typing import List, Tuple, Dict, Set
from operator import itemgetter


class SavingsAlgorithmVRP:
    def __init__(self, depot: Tuple[float, float], customers: List[Tuple[float, float]], 
                 demands: List[int], vehicle_capacity: int, num_vehicles: int):
        """
        Initialize the Clarke-Wright Savings Algorithm VRP solver.
        
        Args:
            depot: (x, y) coordinates of the depot
            customers: List of (x, y) coordinates for each customer
            demands: List of demands for each customer
            vehicle_capacity: Maximum capacity of each vehicle
            num_vehicles: Number of available vehicles
        """
        self.depot = depot
        self.customers = customers
        self.demands = demands
        self.vehicle_capacity = vehicle_capacity
        self.num_vehicles = num_vehicles
        self.num_customers = len(customers)
        
        # Precompute distance matrix
        self.distance_matrix = self._compute_distance_matrix()
        
        # Calculate savings for all customer pairs
        self.savings_list = self._calculate_savings()
    
    def _compute_distance_matrix(self) -> List[List[float]]:
        """Compute the distance matrix between all points (depot + customers)."""
        points = [self.depot] + self.customers
        n = len(points)
        matrix = [[0.0] * n for _ in range(n)]
        
        for i in range(n):
            for j in range(n):
                if i != j:
                    matrix[i][j] = self._euclidean_distance(points[i], points[j])
        
        return matrix
    
    def _euclidean_distance(self, point1: Tuple[float, float], 
                          point2: Tuple[float, float]) -> float:
        """Calculate Euclidean distance between two points."""
        return math.sqrt((point1[0] - point2[0])**2 + (point1[1] - point2[1])**2)
    
    def _calculate_savings(self) -> List[Tuple[int, int, float]]:
        """
        Calculate savings for all customer pairs.
        
        Savings(i,j) = distance(depot,i) + distance(depot,j) - distance(i,j)
        
        Returns:
            List of tuples (customer_i, customer_j, savings_value) sorted by savings
        """
        savings = []
        
        for i in range(self.num_customers):
            for j in range(i + 1, self.num_customers):
                # Savings = d(0,i) + d(0,j) - d(i,j)
                # In distance matrix: depot=0, customer_i=i+1, customer_j=j+1
                depot_to_i = self.distance_matrix[0][i + 1]
                depot_to_j = self.distance_matrix[0][j + 1]
                i_to_j = self.distance_matrix[i + 1][j + 1]
                
                savings_value = depot_to_i + depot_to_j - i_to_j
                savings.append((i, j, savings_value))
        
        # Sort by savings in descending order
        savings.sort(key=itemgetter(2), reverse=True)
        return savings
    
    def solve(self) -> Dict:
        """
        Solve VRP using Clarke-Wright Savings Algorithm.
        
        Returns:
            Dictionary containing routes, total distance, and solution info
        """
        # Initialize: each customer is in its own route
        routes = [[i] for i in range(self.num_customers)]
        route_loads = [self.demands[i] for i in range(self.num_customers)]
        
        # Keep track of which route each customer belongs to
        customer_to_route = {i: i for i in range(self.num_customers)}
        
        # Process savings in decreasing order
        for customer_i, customer_j, savings_value in self.savings_list:
            if savings_value <= 0:
                break  # No more positive savings
            
            route_i = customer_to_route[customer_i]
            route_j = customer_to_route[customer_j]
            
            # Skip if customers are already in the same route
            if route_i == route_j:
                continue
            
            # Check if routes can be merged
            if self._can_merge_routes(routes[route_i], routes[route_j], 
                                    route_loads[route_i], route_loads[route_j],
                                    customer_i, customer_j):
                
                # Merge routes
                merged_route, merged_load = self._merge_routes(
                    routes[route_i], routes[route_j], customer_i, customer_j
                )
                
                # Update data structures
                routes[route_i] = merged_route
                route_loads[route_i] = merged_load
                
                # Update customer-to-route mapping for all customers in route_j
                for customer in routes[route_j]:
                    customer_to_route[customer] = route_i
                
                # Mark route_j as empty
                routes[route_j] = []
                route_loads[route_j] = 0
        
        # Remove empty routes and calculate total distance
        final_routes = [route for route in routes if route]
        
        # Check vehicle limit
        if len(final_routes) > self.num_vehicles:
            # Sort routes by load and keep only the best ones
            route_info = [(route, self._calculate_route_distance(route)) 
                         for route in final_routes]
            route_info.sort(key=itemgetter(1))  # Sort by distance
            final_routes = [route for route, _ in route_info[:self.num_vehicles]]
        
        total_distance = sum(self._calculate_route_distance(route) for route in final_routes)
        
        # Check which customers are served
        served_customers = set()
        for route in final_routes:
            served_customers.update(route)
        unvisited = [i for i in range(self.num_customers) if i not in served_customers]
        
        return {
            'routes': final_routes,
            'total_distance': total_distance,
            'num_vehicles_used': len(final_routes),
            'unvisited_customers': unvisited,
            'feasible': len(unvisited) == 0
        }
    
    def _can_merge_routes(self, route1: List[int], route2: List[int], 
                         load1: int, load2: int, customer_i: int, customer_j: int) -> bool:
        """
        Check if two routes can be merged.
        
        Routes can be merged if:
        1. Combined load doesn't exceed vehicle capacity
        2. customer_i and customer_j are at the ends of their respective routes
        """
        # Check capacity constraint
        if load1 + load2 > self.vehicle_capacity:
            return False
        
        # Check if customers are at the ends of their routes
        # customer_i must be at the beginning or end of route1
        # customer_j must be at the beginning or end of route2
        i_at_end = (route1[0] == customer_i or route1[-1] == customer_i)
        j_at_end = (route2[0] == customer_j or route2[-1] == customer_j)
        
        return i_at_end and j_at_end
    
    def _merge_routes(self, route1: List[int], route2: List[int], 
                     customer_i: int, customer_j: int) -> Tuple[List[int], int]:
        """
        Merge two routes connecting customer_i and customer_j.
        
        Returns:
            Tuple of (merged_route, merged_load)
        """
        # Determine the order of merging based on positions of customers
        merged_route = []
        
        # Find positions of customers in their routes
        i_pos = 0 if route1[0] == customer_i else len(route1) - 1
        j_pos = 0 if route2[0] == customer_j else len(route2) - 1
        
        if i_pos == len(route1) - 1 and j_pos == 0:
            # customer_i is at end of route1, customer_j is at start of route2
            merged_route = route1 + route2
        elif i_pos == 0 and j_pos == len(route2) - 1:
            # customer_i is at start of route1, customer_j is at end of route2
            merged_route = route2 + route1
        elif i_pos == len(route1) - 1 and j_pos == len(route2) - 1:
            # Both at the end - reverse route2 and append
            merged_route = route1 + route2[::-1]
        elif i_pos == 0 and j_pos == 0:
            # Both at the beginning - reverse route1 and append route2
            merged_route = route1[::-1] + route2
        
        merged_load = sum(self.demands[customer] for customer in merged_route)
        
        return merged_route, merged_load
    
    def _calculate_route_distance(self, route: List[int]) -> float:
        """Calculate the total distance for a route."""
        if not route:
            return 0.0
        
        distance = 0.0
        
        # Depot to first customer
        distance += self.distance_matrix[0][route[0] + 1]
        
        # Between customers
        for i in range(len(route) - 1):
            distance += self.distance_matrix[route[i] + 1][route[i + 1] + 1]
        
        # Last customer to depot
        distance += self.distance_matrix[route[-1] + 1][0]
        
        return distance
    
    def get_route_info(self, route: List[int]) -> Dict:
        """Get detailed information about a specific route."""
        if not route:
            return {'distance': 0, 'load': 0, 'customers': []}
        
        total_load = sum(self.demands[customer] for customer in route)
        distance = self._calculate_route_distance(route)
        
        return {
            'distance': distance,
            'load': total_load,
            'customers': route,
            'capacity_utilization': total_load / self.vehicle_capacity
        }


def main():
    """Example usage of the Clarke-Wright Savings Algorithm."""
    # Example problem instance
    depot = (0, 0)
    customers = [(4, 4), (6, 2), (8, 6), (2, 8), (10, 4), (12, 2)]
    demands = [3, 5, 2, 4, 6, 3]
    vehicle_capacity = 10
    num_vehicles = 3
    
    # Solve the problem
    solver = SavingsAlgorithmVRP(depot, customers, demands, vehicle_capacity, num_vehicles)
    solution = solver.solve()
    
    # Print results
    print("=== Clarke-Wright Savings Algorithm VRP Solution ===")
    print(f"Total Distance: {solution['total_distance']:.2f}")
    print(f"Vehicles Used: {solution['num_vehicles_used']}/{num_vehicles}")
    print(f"Feasible Solution: {solution['feasible']}")
    
    # Print savings information
    print(f"\nTop 5 Savings:")
    for i, (cust_i, cust_j, savings) in enumerate(solver.savings_list[:5]):
        print(f"  {i+1}. Customers {cust_i}-{cust_j}: Savings = {savings:.2f}")
    
    for i, route in enumerate(solution['routes']):
        route_info = solver.get_route_info(route)
        print(f"\nVehicle {i+1}:")
        print(f"  Route: Depot -> {' -> '.join(map(str, route))} -> Depot")
        print(f"  Distance: {route_info['distance']:.2f}")
        print(f"  Load: {route_info['load']}/{vehicle_capacity}")
        print(f"  Capacity Utilization: {route_info['capacity_utilization']:.1%}")
    
    if solution['unvisited_customers']:
        print(f"\nUnvisited customers: {solution['unvisited_customers']}")


if __name__ == "__main__":
    main()

File: savings_algorithm.R
# Clarke-Wright Savings Algorithm for Vehicle Routing Problem (VRP)
#
# This classical heuristic starts with individual routes from depot to each customer
# and then combines routes based on savings calculations to minimize total distance.

# Function to calculate Euclidean distance between two points
euclidean_distance <- function(point1, point2) {
  sqrt((point1[1] - point2[1])^2 + (point1[2] - point2[2])^2)
}

# Function to compute distance matrix
compute_distance_matrix <- function(depot, customers) {
  points <- rbind(depot, customers)
  n <- nrow(points)
  dist_matrix <- matrix(0, n, n)
  
  for (i in 1:n) {
    for (j in 1:n) {
      if (i != j) {
        dist_matrix[i, j] <- euclidean_distance(points[i, ], points[j, ])
      }
    }
  }
  
  return(dist_matrix)
}

# Function to calculate savings for all customer pairs
calculate_savings <- function(num_customers, distance_matrix) {
  savings_list <- data.frame(
    customer_i = integer(0),
    customer_j = integer(0),
    savings_value = numeric(0)
  )
  
  for (i in 1:num_customers) {
    for (j in (i + 1):num_customers) {
      if (j <= num_customers) {
        # Savings = d(0,i) + d(0,j) - d(i,j)
        # In distance matrix: depot=1, customer_i=i+1, customer_j=j+1
        depot_to_i <- distance_matrix[1, i + 1]
        depot_to_j <- distance_matrix[1, j + 1]
        i_to_j <- distance_matrix[i + 1, j + 1]
        
        savings_value <- depot_to_i + depot_to_j - i_to_j
        
        savings_list <- rbind(savings_list, data.frame(
          customer_i = i,
          customer_j = j,
          savings_value = savings_value
        ))
      }
    }
  }
  
  # Sort by savings in descending order
  savings_list <- savings_list[order(-savings_list$savings_value), ]
  return(savings_list)
}

# Function to check if two routes can be merged
can_merge_routes <- function(route1, route2, demands, vehicle_capacity, customer_i, customer_j) {
  load1 <- sum(demands[route1])
  load2 <- sum(demands[route2])
  
  # Check capacity constraint
  if (load1 + load2 > vehicle_capacity) {
    return(FALSE)
  }
  
  # Check if customers are at the ends of their respective routes
  i_at_end <- (route1[1] == customer_i || route1[length(route1)] == customer_i)
  j_at_end <- (route2[1] == customer_j || route2[length(route2)] == customer_j)
  
  return(i_at_end && j_at_end)
}

# Function to merge two routes
merge_routes <- function(route1, route2, customer_i, customer_j) {
  # Find positions of customers in their routes
  i_pos <- ifelse(route1[1] == customer_i, 1, length(route1))
  j_pos <- ifelse(route2[1] == customer_j, 1, length(route2))
  
  if (i_pos == length(route1) && j_pos == 1) {
    # customer_i is at end of route1, customer_j is at start of route2
    merged_route <- c(route1, route2)
  } else if (i_pos == 1 && j_pos == length(route2)) {
    # customer_i is at start of route1, customer_j is at end of route2
    merged_route <- c(route2, route1)
  } else if (i_pos == length(route1) && j_pos == length(route2)) {
    # Both at the end - reverse route2 and append
    merged_route <- c(route1, rev(route2))
  } else if (i_pos == 1 && j_pos == 1) {
    # Both at the beginning - reverse route1 and append route2
    merged_route <- c(rev(route1), route2)
  } else {
    # This shouldn't happen if can_merge_routes is correct
    merged_route <- c(route1, route2)
  }
  
  return(merged_route)
}

# Function to calculate route distance
calculate_route_distance <- function(route, distance_matrix) {
  if (length(route) == 0) {
    return(0)
  }
  
  distance <- 0
  
  # Depot to first customer
  distance <- distance + distance_matrix[1, route[1] + 1]
  
  # Between customers
  if (length(route) > 1) {
    for (i in 1:(length(route) - 1)) {
      distance <- distance + distance_matrix[route[i] + 1, route[i + 1] + 1]
    }
  }
  
  # Last customer to depot
  distance <- distance + distance_matrix[route[length(route)] + 1, 1]
  
  return(distance)
}

# Main function to solve VRP using Clarke-Wright Savings Algorithm
savings_algorithm_vrp <- function(depot, customers, demands, vehicle_capacity, num_vehicles) {
  num_customers <- nrow(customers)
  distance_matrix <- compute_distance_matrix(depot, customers)
  savings_list <- calculate_savings(num_customers, distance_matrix)
  
  # Initialize: each customer is in its own route
  routes <- lapply(1:num_customers, function(i) i)
  
  # Keep track of which route each customer belongs to
  customer_to_route <- 1:num_customers
  
  # Process savings in decreasing order
  for (row_idx in 1:nrow(savings_list)) {
    savings_row <- savings_list[row_idx, ]
    customer_i <- savings_row$customer_i
    customer_j <- savings_row$customer_j
    savings_value <- savings_row$savings_value
    
    if (savings_value <= 0) {
      break  # No more positive savings
    }
    
    route_i <- customer_to_route[customer_i]
    route_j <- customer_to_route[customer_j]
    
    # Skip if customers are already in the same route
    if (route_i == route_j) {
      next
    }
    
    # Check if routes can be merged
    if (can_merge_routes(routes[[route_i]], routes[[route_j]], demands, vehicle_capacity, customer_i, customer_j)) {
      # Merge routes
      merged_route <- merge_routes(routes[[route_i]], routes[[route_j]], customer_i, customer_j)
      
      # Update data structures
      routes[[route_i]] <- merged_route
      
      # Update customer-to-route mapping for all customers in route_j
      for (customer in routes[[route_j]]) {
        customer_to_route[customer] <- route_i
      }
      
      # Mark route_j as empty
      routes[[route_j]] <- integer(0)
    }
  }
  
  # Remove empty routes
  final_routes <- routes[sapply(routes, length) > 0]
  
  # Check vehicle limit
  if (length(final_routes) > num_vehicles) {
    # Sort routes by distance and keep only the best ones
    route_distances <- sapply(final_routes, function(route) calculate_route_distance(route, distance_matrix))
    route_order <- order(route_distances)
    final_routes <- final_routes[route_order[1:num_vehicles]]
  }
  
  total_distance <- sum(sapply(final_routes, function(route) calculate_route_distance(route, distance_matrix)))
  
  # Check which customers are served
  served_customers <- unlist(final_routes)
  unvisited <- setdiff(1:num_customers, served_customers)
  
  return(list(
    routes = final_routes,
    total_distance = total_distance,
    num_vehicles_used = length(final_routes),
    unvisited_customers = unvisited,
    feasible = length(unvisited) == 0,
    savings_list = savings_list
  ))
}

# Function to get detailed route information
get_route_info <- function(route, demands, vehicle_capacity, distance_matrix) {
  if (length(route) == 0) {
    return(list(distance = 0, load = 0, customers = integer(0)))
  }
  
  total_load <- sum(demands[route])
  distance <- calculate_route_distance(route, distance_matrix)
  
  return(list(
    distance = distance,
    load = total_load,
    customers = route,
    capacity_utilization = total_load / vehicle_capacity
  ))
}

# Example usage function
run_savings_algorithm_example <- function() {
  cat("=== Clarke-Wright Savings Algorithm VRP Solution ===\n")
  
  # Example problem instance
  depot <- c(0, 0)
  customers <- matrix(c(
    4, 4,
    6, 2,
    8, 6,
    2, 8,
    10, 4,
    12, 2
  ), ncol = 2, byrow = TRUE)
  
  demands <- c(3, 5, 2, 4, 6, 3)
  vehicle_capacity <- 10
  num_vehicles <- 3
  
  # Solve the problem
  solution <- savings_algorithm_vrp(depot, customers, demands, vehicle_capacity, num_vehicles)
  distance_matrix <- compute_distance_matrix(depot, customers)
  
  # Print results
  cat(sprintf("Total Distance: %.2f\n", solution$total_distance))
  cat(sprintf("Vehicles Used: %d/%d\n", solution$num_vehicles_used, num_vehicles))
  cat(sprintf("Feasible Solution: %s\n", solution$feasible))
  
  # Print top savings
  cat("\nTop 5 Savings:\n")
  top_savings <- head(solution$savings_list, 5)
  for (i in 1:nrow(top_savings)) {
    cat(sprintf("  %d. Customers %d-%d: Savings = %.2f\n", 
                i, top_savings[i, ]$customer_i, top_savings[i, ]$customer_j, top_savings[i, ]$savings_value))
  }
  
  # Print detailed routes
  for (i in seq_along(solution$routes)) {
    route <- solution$routes[[i]]
    route_info <- get_route_info(route, demands, vehicle_capacity, distance_matrix)
    
    cat(sprintf("\nVehicle %d:\n", i))
    route_str <- paste(route, collapse = " -> ")
    cat(sprintf("  Route: Depot -> %s -> Depot\n", route_str))
    cat(sprintf("  Distance: %.2f\n", route_info$distance))
    cat(sprintf("  Load: %d/%d\n", route_info$load, vehicle_capacity))
    cat(sprintf("  Capacity Utilization: %.1f%%\n", route_info$capacity_utilization * 100))
  }
  
  if (length(solution$unvisited_customers) > 0) {
    unvisited_str <- paste(solution$unvisited_customers, collapse = ", ")
    cat(sprintf("\nUnvisited customers: %s\n", unvisited_str))
  }
  
  return(solution)
}

# Run example if script is executed directly
if (!interactive()) {
  run_savings_algorithm_example()
}